File: ios/Classes/Bridges/ConfigurationStatusBridge.swift
import Flutter
import SuperwallKit

// Abstract base class for subscription status enum
public class ConfigurationStatusBridge: BridgeInstance {
  var configurationStatus: ConfigurationStatus {
    assertionFailure("Subclasses must implement")
    return .pending
  }
}

public class ConfigurationStatusPendingBridge: ConfigurationStatusBridge {
  class override var bridgeClass: BridgeClass { "ConfigurationStatusPendingBridge" }
  override var configurationStatus: ConfigurationStatus { .pending }
}

public class ConfigurationStatusConfiguredBridge: ConfigurationStatusBridge {
  class override var bridgeClass: BridgeClass { "ConfigurationStatusConfiguredBridge" }
  override var configurationStatus: ConfigurationStatus { .configured }
}

public class ConfigurationStatusFailedBridge: ConfigurationStatusBridge {
  class override var bridgeClass: BridgeClass { "ConfigurationStatusFailedBridge" }
  override var configurationStatus: ConfigurationStatus { .failed }
}

extension ConfigurationStatus {
  func createBridgeId() -> BridgeId {
    switch self {
      case .pending:
        let bridgeInstance = BridgingCreator.shared.createBridgeInstance(bridgeClass: ConfigurationStatusPendingBridge.bridgeClass)
        return bridgeInstance.bridgeId
      case .configured:
        let bridgeInstance = BridgingCreator.shared.createBridgeInstance(bridgeClass: ConfigurationStatusConfiguredBridge.bridgeClass)
        return bridgeInstance.bridgeId
      case .failed:
        let bridgeInstance = BridgingCreator.shared.createBridgeInstance(bridgeClass: ConfigurationStatusFailedBridge.bridgeClass)
        return bridgeInstance.bridgeId
    }
  }
}

File: android/src/main/kotlin/com/superwall/superwallkit_flutter/bridges/ConfigurationStatusBridge.kt
package com.superwall.superwallkit_flutter.bridges

import com.superwall.sdk.config.models.ConfigurationStatus
import android.content.Context
import com.superwall.superwallkit_flutter.BridgingCreator

abstract class ConfigurationStatusBridge(context: Context, bridgeId: BridgeId) : BridgeInstance(context,
    bridgeId
) {
    abstract val configurationStatus: ConfigurationStatus
}

class ConfigurationStatusPendingBridge(context: Context, bridgeId: BridgeId) : ConfigurationStatusBridge(context,
    bridgeId
) {


    companion object {
        fun bridgeClass(): BridgeClass = "ConfigurationStatusPendingBridge"
    }

    override val configurationStatus: ConfigurationStatus
        get() = ConfigurationStatus.Configured
}


class ConfigurationStatusFailedBridge(context: Context, bridgeId: BridgeId) : ConfigurationStatusBridge(context,
    bridgeId
) {


    companion object {
        fun bridgeClass(): BridgeClass = "ConfigurationStatusFailedBridge"
    }

    override val configurationStatus: ConfigurationStatus
        get() = ConfigurationStatus.Failed
}

class ConfigurationStatusConfiguredBridge(context: Context, bridgeId: BridgeId) : ConfigurationStatusBridge(context,
    bridgeId
) {


    companion object {
        fun bridgeClass(): BridgeClass = "ConfigurationStatusConfiguredBridge"
    }

    override val configurationStatus: ConfigurationStatus
        get() = ConfigurationStatus.Configured
}


suspend fun ConfigurationStatus.createBridgeId(): BridgeId {

    val bridgeClass = when (this) {
        is ConfigurationStatus.Pending -> ConfigurationStatusPendingBridge.bridgeClass()
        is ConfigurationStatus.Failed -> ConfigurationStatusFailedBridge.bridgeClass()
        is ConfigurationStatus.Configured -> ConfigurationStatusConfiguredBridge.bridgeClass()
    }
    val bridge = BridgingCreator.shared().createBridgeInstanceFromBridgeClass(
        bridgeClass = bridgeClass
    )
    return bridge.bridgeId
}